### PR TITLE
fastjet/internal/heap: implement a min heap for pairs of jets

### PR DESCRIPTION
**Pull request by Bastiantheone**
This PR implements a min heap that will be used in the delaunay clustering.

### CODE REVIEW

File: fastjet/internal/heap/heap.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
+// Copyright 2017 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package heap
+
+// pair is an item in the heap. It contains the indices the two jets to be
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: It contains the indices *of* the two...
Right?

File: fastjet/internal/heap/heap.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+type Heap struct {
+	n     int // n is the last index of the slice
+	items []pair
+}
+
+// NewHeap returns a heap pointer.
+func NewHeap() *Heap {
+	h := &Heap{n: 0}
+	// pair{} is a placeholder at h.items[0]
+	h.items = append(h.items, pair{})
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: You could just as well write 'items: make([]pair, 1)' when creating 'h' and be done with it, right?

File: fastjet/internal/heap/heap.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// Heap contains a slice of items and the last index.
+type Heap struct {
+	n     int // n is the last index of the slice
+	items []pair
+}
+
+// NewHeap returns a heap pointer.
+func NewHeap() *Heap {
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: Perhaps simply name it 'New' ?

File: fastjet/internal/heap/heap.go (hunk 1 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: yes fixed

File: fastjet/internal/heap/heap.go (hunk 2 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: yes it's fixed

File: fastjet/internal/heap/heap.go (hunk 3 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: done

File: fastjet/internal/heap/heap.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	return item.jeti, item.jetj, item.dij
+}
+
+// IsEmpty returns whether a heap is empty.
+func (h *Heap) IsEmpty() bool {
+	return h.n == 0
+}
+
+// moveUp compares an item with its parent and moves it up if it has
+// a lower distance. It will keep moving up until the parent's distance is less
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: `s/is less/is smaller/`

File: fastjet/internal/heap/heap.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+// or it reaches the top.
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: `s/or it/or if it/`

File: fastjet/internal/heap/heap.go (hunk 4 above)
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: `s/a lower/a smaller/`

File: fastjet/internal/heap/heap.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		return
+	}
+	if h.less(i, parent) {
+		h.swap(i, parent)
+		h.moveUp(parent)
+	}
+}
+
+// moveDown compares an item to its children and moves it down
+// if one of the children has a lower distance. It will keep
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: `s/lower/smaller/`

File: fastjet/internal/heap/heap.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+// moving down until it reaches a leaf or both children have
+// a higher distance.
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: `s/higher/bigger/`

File: fastjet/internal/heap/heap.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	} else {
+		smallestChild = rightChild
+	}
+	if h.less(smallestChild, i) {
+		h.swap(i, smallestChild)
+		h.moveDown(smallestChild)
+	}
+}
+
+// less returns whether item at index i has a lower distance
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: `s/lower/smaller/`

File: fastjet/internal/heap/heap.go (hunk 4 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: done

File: fastjet/internal/heap/heap.go (hunk 5 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: done

File: fastjet/internal/heap/heap.go (hunk 4 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: done

File: fastjet/internal/heap/heap.go (hunk 6 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: done

File: fastjet/internal/heap/heap.go (hunk 7 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: done

File: fastjet/internal/heap/heap.go (hunk 8 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: done

**Comment by sbinet** [comment]
one thing I forgot to mention: how does this compare (performance-wise) with `container/heap`:
https://godoc.org/container/heap

(presumably, we should be faster, right?, because `interface{}`, but... also: do they give the same answer :P)

**Comment by Bastiantheone** [comment]
They give the same answer for the test I ran. I also ran benchmarks for pushing 500 random items and then poping them.
For the container/heap the numbers are:
`5000	    320137 ns/op	    4104 B/op	      10 allocs/op`
For my version:
`10000	    190855 ns/op	   16368 B/op	      10 allocs/op`

**Comment by sbinet** [comment]
ok. LGTM (the increase in terms of number of bytes allocated /op is a bit surprising though...)

**Comment by Bastiantheone** [comment]
A potential cause is, that my up and down are called recursively, while container/heap handles everything in a for-loop.

**Comment by Bastiantheone** [comment]
Nevermind on that

File: fastjet/internal/heap/heap.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+// a bigger distance.
+func (h *Heap) moveDown(i int) {
+	for {
+		leftChid := 2 * i
+		if leftChid > h.n {
+			return
+		}
+		rightChild := 2*i + 1
+		var smallestChild int
+		if leftChid == h.n {
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: replace with a `switch` ?

```go
switch {
case leftChild == h.n:
  // ...
case h.less(leftChild, rightChild):
  // ...
default:
  // ...
}
```

File: fastjet/internal/heap/heap.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	}
+}
+
+// moveDown compares an item to its children and moves it down
+// if one of the children has a smaller distance. It will keep
+// moving down until it reaches a leaf or both children have
+// a bigger distance.
+func (h *Heap) moveDown(i int) {
+	for {
+		leftChid := 2 * i
```
Comment (sbinet) [inline fastjet/internal/heap/heap.go]: `s/leftChid/leftChild/`

Comment (sbinet) [inline fastjet/internal/heap/heap.go]: or even:
`left` and, similarly, `right` ?

File: fastjet/internal/heap/heap.go (hunk 9 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: done

File: fastjet/internal/heap/heap.go (hunk 10 above)
Comment (Bastiantheone) [inline fastjet/internal/heap/heap.go]: done

**Comment by Bastiantheone** [comment]
Done

**Comment by sbinet** [comment]
thanks

**Pull request merged**